src/Facility.CodeGen.CSharp/CSharpGenerator.cs

```diff
@@ -356,11 +375,80 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
 									}
 								}
 
-								if (validatableFields.Count != 0)
+								if (validateFields.Count != 0)
+								{
+									code.WriteLineSkipOnce();
+									foreach (var fieldInfo in validateFields)
+									{
+										var type = context.GetFieldType(fieldInfo);
+										var propertyName = context.GetFieldPropertyName(fieldInfo);
+										var validation = fieldInfo.Validation!;
+
+										switch (type.Kind)
+										{
+											case ServiceTypeKind.Enum:
+											{
+												code.WriteLine($"if (!{propertyName}.IsDefined())");
+												using (code.Indent())
+													code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{propertyName}\");");
+
+												break;
+											}
+
+											case ServiceTypeKind.String:
+											{
+												var validRange = validation.LengthRange;
+												if (validRange != null)
+													GenerateRangeCheck(code, $"{propertyName}.Length", validRange);
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													code.WriteLine($"if (!System.Text.RegularExpressions.Regex.IsMatch({propertyName}, \"{validPattern}\")");
```

Comment:
The pattern needs to be escaped. Use `CSharpUtility.CreateString`.

Also, we shouldn't rely on the [global regex cache](https://docs.microsoft.com/en-us/dotnet/standard/base-types/compilation-and-reuse-in-regular-expressions#the-regular-expressions-cache), which means we should create a `static readonly Regex` for each pattern.

Also, feel free to add `System.Text.RegularExpressions` to the usings.

Reply:
I'm curious. Does the cache involve some serious overhead in practice? Or is it just that LRU evictions could cause sneaky (and seriously non-deterministic) performance issues?

FWIW, My original thought was to avoid unused using warnings in generated code, assuming that kind of warning was something generated code should be concerned about. What are your thoughts on generated code style? Is the bar just generating reasonable looking code, or are there specifics beyond that?

Reply:
> LRU evictions could cause sneaky (and seriously non-deterministic) performance issues

Yep.

The generated code already has unused usings, so you wouldn't be adding a new issue. Generated code should be exempt from code style warnings.

I like generating code that's as normal-looking as possible.